core/kernel/api/class.ModelFactory.php

```diff
@@ -16,112 +17,143 @@
+    }
+
     /**
      * @author "Lionel Lecaque, <lionel@taotesting.com>"
+     *
      * @param string $namespace
+     *
      * @return string
      */
-    private function getModelId($namespace){
-        $dbWrapper = core_kernel_classes_DbWrapper::singleton();
-        
+    public function getModelId($namespace)
+    {
         $query = 'SELECT modelid FROM models WHERE (modeluri = ?)';
-        $results = $dbWrapper->query($query, array($namespace));
-       
-        return $results->fetchColumn(0);
+        $results = $this->dbWrapper->query($query, [$namespace]);
 
+        return $results->fetchColumn(0);
     }
-    
+
     /**
      * @author "Lionel Lecaque, <lionel@taotesting.com>"
+     *
      * @param string $namespace
+     *
+     * @return string new added model id
      */
-    private function addNewModel($namespace){
-        $dbWrapper = core_kernel_classes_DbWrapper::singleton();
-        $results = $dbWrapper->insert('models', array('modeluri' =>$namespace));
-        
-        
+    public function addNewModel($namespace)
+    {
+        $modelId = $this->getUniquePrimaryKey();
+        $this->dbWrapper->insert('models', ['modelid' => $modelId, 'modeluri' => $namespace]);
+
+        return $modelId;
     }
-    
+
     /**
      * @author "Lionel Lecaque, <lionel@taotesting.com>"
+     *
      * @param string $namespace
      * @param string $data xml content
+     * @return bool Were triples added?
      */
-    public function createModel($namespace, $data){
-
+    public function createModel($namespace, $data)
+    {
         $modelId = $this->getModelId($namespace);
-        if($modelId === false){
-            common_Logger::d('modelId not found, need to add namespace '. $namespace);
-            $this->addNewModel($namespace);
-            //TODO bad way, need to find better
-            $modelId = $this->getModelId($namespace);
+        if ($modelId === false) {
+            common_Logger::d('modelId not found, need to add namespace ' . $namespace);
+            $modelId = $this->addNewModel($namespace);
         }
         $modelDefinition = new EasyRdf_Graph($namespace);
-        if(is_file($data)){
+        if (is_file($data)) {
             $modelDefinition->parseFile($data);
-        }else {
+        } else {
             $modelDefinition->parse($data);
         }
-        $graph = $modelDefinition->toRdfPhp();
-        $resources = $modelDefinition->resources();
         $format = EasyRdf_Format::getFormat('php');
-        
+
         $data = $modelDefinition->serialise($format);
-        
-        foreach ($data as $subjectUri => $propertiesValues){
-            foreach ($propertiesValues as $prop=>$values){
-                foreach ($values as $k => $v) {
-                    $this->addStatement($modelId, $subjectUri, $prop, $v['value'], isset($v['lang']) ? $v['lang'] : null);
+
+        $returnValue = false;
+
+        foreach ($data as $subjectUri => $propertiesValues) {
```

Comment:
how about removing the 3rd foreach 
```
 foreach ($data as $subjectUri => $propertiesValues) {
            foreach ($propertiesValues as $prop => $values) {
                $v = current($values);
                $returnValue |= $this->addStatement($modelId, $subjectUri, $prop, $v['value'],
                    isset($v['lang']) ? $v['lang'] : null);
            }
        }
```
this only if we have only one array inside the $values